src/KOKKOS/metatensor_system_kokkos.cpp

```diff
@@ -0,0 +1,412 @@
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing authors: Guillaume Fraux <guillaume.fraux@epfl.ch>
+------------------------------------------------------------------------- */
+#include "metatensor_system_kokkos.h"
+
+#include "atom.h"
+#include "domain.h"
+#include "error.h"
+#include "neighbor.h"
+
+#include "neigh_list.h"
+#include "neigh_request.h"
+
+#include "kokkos.h"
+#include "atom_kokkos.h"
+
+#include <torch/cuda.h>
+#include <chrono>
+
+#ifndef KOKKOS_ENABLE_CUDA
+// fake Kokkos::Cuda for non-CUDA builds
+namespace Kokkos {
+class Cuda {};
+} // namespace Kokkos
+#endif // KOKKOS_ENABLE_CUDA
+
+using namespace LAMMPS_NS;
+
+/* ---------------------------------------------------------------------- */
+
+template<class DeviceType>
+MetatensorSystemAdaptorKokkos<DeviceType>::MetatensorSystemAdaptorKokkos(LAMMPS *lmp, Pair* requestor, MetatensorSystemOptionsKokkos<DeviceType> options):
+    Pointers(lmp),
+    list_(nullptr),
+    options_(std::move(options)),
+    caches_(),
+    atomic_types_(torch::zeros({0}, torch::TensorOptions().dtype(torch::kInt32)))
+{
+    // We ask LAMMPS for a full neighbor lists because we need to know about
+    // ALL pairs, even if options->full_list() is false. We will then filter
+    // the pairs to only include each pair once where needed.
+    auto request = neighbor->add_request(requestor, NeighConst::REQ_FULL | NeighConst::REQ_GHOST);
+    request->set_id(0);
+    request->set_cutoff(options_.interaction_range);
+    // set whether the kokkos NL should be calculated on host or device
+    request->set_kokkos_host(std::is_same_v<DeviceType,LMPHostType> &&
+                            !std::is_same_v<DeviceType,LMPDeviceType>);
+    request->set_kokkos_device(std::is_same_v<DeviceType,LMPDeviceType>);
+}
+
+template<class DeviceType>
+MetatensorSystemAdaptorKokkos<DeviceType>::~MetatensorSystemAdaptorKokkos() {
+
+}
+
+template<class DeviceType>
+void MetatensorSystemAdaptorKokkos<DeviceType>::init_list(int id, NeighList* ptr) {
+    assert(id == 0);
+    list_ = ptr;
+}
+
+template<class DeviceType>
+void MetatensorSystemAdaptorKokkos<DeviceType>::add_nl_request(double cutoff, metatensor_torch::NeighborListOptions request) {
+    if (cutoff > options_.interaction_range) {
+        error->all(FLERR,
+            "Invalid metatensor model: one of the requested neighbor lists "
+            "has a cutoff ({}) larger than the model interaction range ({})",
+            cutoff, options_.interaction_range
+        );
+    } else if (cutoff < 0 || !std::isfinite(cutoff)) {
+        error->all(FLERR,
+            "model requested an invalid cutoff for neighbors list: {} "
+            "(cutoff in model units is {})",
+            cutoff, request->cutoff()
+        );
+    }
```

Comment:
It would be good to do such checks in a single common place for both the kokkos & non-kokkos builds. Maybe a separate base class?